File: cpp/src/IceWS/EndpointI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2014 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <IceWS/EndpointI.h>
#include <IceWS/AcceptorI.h>
#include <IceWS/ConnectorI.h>
#include <IceWS/TransceiverI.h>
#include <IceWS/Instance.h>
#include <Ice/BasicStream.h>
#include <Ice/LocalException.h>
#include <Ice/IPEndpointI.h>
#include <Ice/HashUtil.h>

using namespace std;
using namespace Ice;
using namespace IceWS;

IceWS::EndpointI::EndpointI(const InstancePtr& instance, const IceInternal::EndpointIPtr& del, const string& res) :
    _instance(instance), _delegate(IceInternal::IPEndpointIPtr::dynamicCast(del)), _resource(res)
{
}

IceWS::EndpointI::EndpointI(const InstancePtr& instance, const IceInternal::EndpointIPtr& del, vector<string>& args) :
    _instance(instance), _delegate(IceInternal::IPEndpointIPtr::dynamicCast(del))
{
    initWithOptions(args);

    if(_resource.empty())
    {
        const_cast<string&>(_resource) = "/";
    }
}

IceWS::EndpointI::EndpointI(const InstancePtr& instance, const IceInternal::EndpointIPtr& del, 
                            IceInternal::BasicStream* s) :
    _instance(instance), _delegate(IceInternal::IPEndpointIPtr::dynamicCast(del))
{
    s->read(const_cast<string&>(_resource), false);
}

Ice::EndpointInfoPtr
IceWS::EndpointI::getInfo() const
{
    class InfoI : public IceWS::EndpointInfo
    {
    public:

        InfoI(const IceInternal::EndpointIPtr& e) : _endpoint(e)
        {
        }

        virtual Short
        type() const
        {
            return _endpoint->type();
        }

        virtual bool
        datagram() const
        {
            return _endpoint->datagram();
        }

        virtual bool
        secure() const
        {
            return _endpoint->secure();
        }
        
    private:
        
        const IceInternal::EndpointIPtr _endpoint;
    };

    EndpointInfoPtr info = new InfoI(const_cast<EndpointI*>(this));
    info->timeout = _delegate->timeout();
    info->compress = _delegate->compress();
    _delegate->fillEndpointInfo(info.get());
    info->resource = _resource;
    return info;
}

Ice::Short
IceWS::EndpointI::type() const
{
    return _delegate->type();
}

const string&
IceWS::EndpointI::protocol() const
{
    return _delegate->protocol();
}

void
IceWS::EndpointI::streamWrite(IceInternal::BasicStream* s) const
{
    s->startWriteEncaps();
    _delegate->streamWriteImpl(s);
    s->write(_resource, false);
    s->endWriteEncaps();
}

Int
IceWS::EndpointI::timeout() const
{
    return _delegate->timeout();
}

IceInternal::EndpointIPtr
IceWS::EndpointI::timeout(Int timeout) const
{
    if(timeout == _delegate->timeout())
    {
        return const_cast<EndpointI*>(this);
    }
    else
    {
        return new EndpointI(_instance, _delegate->timeout(timeout), _resource);
    }
}

const string&
IceWS::EndpointI::connectionId() const
{
    return _delegate->connectionId();
}

IceInternal::EndpointIPtr
IceWS::EndpointI::connectionId(const string& connectionId) const
{
    if(connectionId == _delegate->connectionId())
    {
        return const_cast<EndpointI*>(this);
    }
    else
    {
        return new EndpointI(_instance, _delegate->connectionId(connectionId), _resource);
    }
}

bool
IceWS::EndpointI::compress() const
{
    return _delegate->compress();
}

IceInternal::EndpointIPtr
IceWS::EndpointI::compress(bool compress) const
{
    if(compress == _delegate->compress())
    {
        return const_cast<EndpointI*>(this);
    }
    else
    {
        return new EndpointI(_instance, _delegate->compress(compress), _resource);
    }
}

bool
IceWS::EndpointI::datagram() const
{
    return _delegate->datagram();
}

bool
IceWS::EndpointI::secure() const
{
    return _delegate->secure();
}

IceInternal::TransceiverPtr
IceWS::EndpointI::transceiver(IceInternal::EndpointIPtr& endp) const
{
    endp = const_cast<EndpointI*>(this);
    return 0;
}

vector<IceInternal::ConnectorPtr>
IceWS::EndpointI::connectors(Ice::EndpointSelectionType selType) const
{
    vector<IceInternal::ConnectorPtr> connectors = _delegate->connectors(selType);
    for(vector<IceInternal::ConnectorPtr>::iterator p = connectors.begin(); p != connectors.end(); ++p)
    {
        *p = new ConnectorI(_instance, *p, _delegate->host(), _delegate->port(), _resource);
    }
    return connectors;
}

void
IceWS::EndpointI::connectors_async(Ice::EndpointSelectionType selType,
                                   const IceInternal::EndpointI_connectorsPtr& callback) const
{
    class CallbackI : public IceInternal::EndpointI_connectors
    {
    public:

        CallbackI(const IceInternal::EndpointI_connectorsPtr& callback, const InstancePtr& instance, 
                  const string& host, int port, const string& resource) : 
            _callback(callback), _instance(instance), _host(host), _port(port), _resource(resource)
        {
        }

        virtual void connectors(const vector<IceInternal::ConnectorPtr>& c)
        {
            vector<IceInternal::ConnectorPtr> connectors = c;
            for(vector<IceInternal::ConnectorPtr>::iterator p = connectors.begin(); p != connectors.end(); ++p)
            {
                *p = new ConnectorI(_instance, *p, _host, _port, _resource);
            }
            _callback->connectors(connectors);
        }

        virtual void exception(const Ice::LocalException& ex)
        {
            _callback->exception(ex);
        }

    private:
        
        const IceInternal::EndpointI_connectorsPtr _callback;
        const InstancePtr _instance;
        const string _host;
        const int _port;
        const string _resource;
    };
    _delegate->connectors_async(selType, new CallbackI(callback, _instance, _delegate->host(), _delegate->port(), 
                                                       _resource));
}

IceInternal::AcceptorPtr
IceWS::EndpointI::acceptor(IceInternal::EndpointIPtr& endp, const string& adapterName) const
{
    IceInternal::EndpointIPtr delEndp;
    IceInternal::AcceptorPtr delAcc = _delegate->acceptor(delEndp, adapterName);
    if(delEndp)
    {
        endp = new EndpointI(_instance, delEndp, _resource);
    }
    return new AcceptorI(_instance, delAcc);
}

vector<IceInternal::EndpointIPtr>
IceWS::EndpointI::expand() const
{
    vector<IceInternal::EndpointIPtr> endps = _delegate->expand();
    for(vector<IceInternal::EndpointIPtr>::iterator p = endps.begin(); p != endps.end(); ++p)
    {
        *p = p->get() == _delegate.get() ? const_cast<EndpointI*>(this) : new EndpointI(_instance, *p, _resource);
    }
    return endps;
}

bool
IceWS::EndpointI::equivalent(const IceInternal::EndpointIPtr& endpoint) const
{
    const EndpointI* wsEndpointI = dynamic_cast<const EndpointI*>(endpoint.get());
    if(!wsEndpointI)
    {
        return false;
    }
    return _delegate->equivalent(wsEndpointI->_delegate);
}

Ice::Int
IceWS::EndpointI::hash() const
{
    int h = _delegate->hash();
    IceInternal::hashAdd(h, _resource);
    return h;
}

string
IceWS::EndpointI::options() const
{
    //
    // WARNING: Certain features, such as proxy validation in Glacier2,
    // depend on the format of proxy strings. Changes to toString() and
    // methods called to generate parts of the reference string could break
    // these features. Please review for all features that depend on the
    // format of proxyToString() before changing this and related code.
    //
    ostringstream s;
    s << _delegate->options();

    if(!_resource.empty())
    {
        s << " -r ";
        bool addQuote = _resource.find(':') != string::npos;
        if(addQuote)
        {
            s << "\"";
        }
        s << _resource;
        if(addQuote)
        {
            s << "\"";
        }
    }

    return s.str();
}

bool
IceWS::EndpointI::operator==(const Ice::LocalObject& r) const
{
    const EndpointI* p = dynamic_cast<const EndpointI*>(&r);
    if(!p)
    {
        return false;
    }

    if(this == p)
    {
        return true;
    }

    if(_delegate != p->_delegate)
    {
        return false;
    }

    if(_resource != p->_resource)
    {
        return false;
    }

    return true;
}

bool
IceWS::EndpointI::operator<(const Ice::LocalObject& r) const
{
    const EndpointI* p = dynamic_cast<const EndpointI*>(&r);
    if(!p)
    {
        const IceInternal::EndpointI* e = dynamic_cast<const IceInternal::EndpointI*>(&r);
        if(!e)
        {
            return false;
        }
        return type() < e->type();
    }

    if(this == p)
    {
        return false;
    }

    if(_delegate < p->_delegate)
    {
        return true;
    }
    else if (p->_delegate < _delegate)
    {
        return false;
    }

    if(_resource < p->_resource)
    {
        return true;
    }
    else if (p->_resource < _resource)
    {
        return false;
    }

    return false;

}    

bool
IceWS::EndpointI::checkOption(const string& option, const string& argument, const string& endpoint)
{
    switch(option[1])
    {
    case 'r':
    {
        if(argument.empty())
        {
            EndpointParseException ex(__FILE__, __LINE__);
            ex.str = "no argument provided for -r option in endpoint " + endpoint + _delegate->options();
            throw ex;
        }
        const_cast<string&>(_resource) = argument;
        return true;
    }

    default:
    {
        return false;
    }
    }
}

IceWS::EndpointFactoryI::EndpointFactoryI(const InstancePtr& instance, const IceInternal::EndpointFactoryPtr& del) :
    _instance(instance),
    _delegate(del)
{
}

IceWS::EndpointFactoryI::~EndpointFactoryI()
{
}

Short
IceWS::EndpointFactoryI::type() const
{
    return _instance->type();
}

string
IceWS::EndpointFactoryI::protocol() const
{
    return _instance->protocol();
}

IceInternal::EndpointIPtr
IceWS::EndpointFactoryI::create(vector<string>& args, bool oaEndpoint) const
{
    return new EndpointI(_instance, _delegate->create(args, oaEndpoint), args);
}

IceInternal::EndpointIPtr
IceWS::EndpointFactoryI::read(IceInternal::BasicStream* s) const
{
    return new EndpointI(_instance, _delegate->read(s), s);
}

void
IceWS::EndpointFactoryI::destroy()
{
    _delegate->destroy();
    _instance = 0;
}

IceInternal::EndpointFactoryPtr
IceWS::EndpointFactoryI::clone(const IceInternal::ProtocolInstancePtr&) const
{
    assert(false); // We don't support cloning this transport.
    return 0;
}

File: java/src/IceWS/EndpointI.java
// **********************************************************************
//
// Copyright (c) 2003-2014 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceWS;

final class EndpointI extends IceInternal.EndpointI
{
    public EndpointI(Instance instance, IceInternal.EndpointI del, String res)
    {
        _instance = instance;
        _delegate = (IceInternal.IPEndpointI)del;
        _resource = res;
    }

    public EndpointI(Instance instance, IceInternal.EndpointI del, java.util.ArrayList<String> args)
    {
        _instance = instance;
        _delegate = (IceInternal.IPEndpointI)del;

        initWithOptions(args);

        if(_resource == null)
        {
            _resource = "/";
        }
    }

    public EndpointI(Instance instance, IceInternal.EndpointI del, IceInternal.BasicStream s)
    {
        _instance = instance;
        _delegate = (IceInternal.IPEndpointI)del;

        _resource = s.readString();
    }

    @Override
    public Ice.EndpointInfo getInfo()
    {
        IceWS.EndpointInfo info = new IceWS.EndpointInfo()
            {
                @Override
                public short type()
                {
                    return EndpointI.this.type();
                }

                @Override
                public boolean datagram()
                {
                    return EndpointI.this.datagram();
                }

                @Override
                public boolean secure()
                {
                    return EndpointI.this.secure();
                }
            };

        info.timeout = _delegate.timeout();
        info.compress = _delegate.compress();
        _delegate.fillEndpointInfo(info);
        info.resource = _resource;
        return info;
    }

    @Override
    public short type()
    {
        return _delegate.type();
    }

    @Override
    public String protocol()
    {
        return _delegate.protocol();
    }

    @Override
    public void streamWrite(IceInternal.BasicStream s)
    {
        s.startWriteEncaps();
        _delegate.streamWriteImpl(s);
        s.writeString(_resource);
        s.endWriteEncaps();
    }

    @Override
    public int timeout()
    {
        return _delegate.timeout();
    }

    @Override
    public IceInternal.EndpointI timeout(int timeout)
    {
        if(timeout == _delegate.timeout())
        {
            return this;
        }
        else
        {
            return new EndpointI(_instance, _delegate.timeout(timeout), _resource);
        }
    }

    @Override
    public String connectionId()
    {
        return _delegate.connectionId();
    }

    @Override
    public IceInternal.EndpointI connectionId(String connectionId)
    {
        if(connectionId.equals(_delegate.connectionId()))
        {
            return this;
        }
        else
        {
            return new EndpointI(_instance, _delegate.connectionId(connectionId), _resource);
        }
    }

    @Override
    public boolean compress()
    {
        return _delegate.compress();
    }

    @Override
    public IceInternal.EndpointI compress(boolean compress)
    {
        if(compress == _delegate.compress())
        {
            return this;
        }
        else
        {
            return new EndpointI(_instance, _delegate.compress(compress), _resource);
        }
    }

    @Override
    public boolean datagram()
    {
        return _delegate.datagram();
    }

    @Override
    public boolean secure()
    {
        return _delegate.secure();
    }

    @Override
    public IceInternal.Transceiver transceiver(IceInternal.EndpointIHolder endpoint)
    {
        endpoint.value = this;
        return null;
    }

    @Override
    public java.util.List<IceInternal.Connector> connectors(Ice.EndpointSelectionType selType)
    {
        java.util.List<IceInternal.Connector> connectors = _delegate.connectors(selType);
        java.util.List<IceInternal.Connector> l = new java.util.ArrayList<IceInternal.Connector>();
        for(IceInternal.Connector c : connectors)
        {
            l.add(new ConnectorI(_instance, c, _delegate.host(), _delegate.port(), _resource));
        }
        return l;
    }

    @Override
    public void connectors_async(Ice.EndpointSelectionType selType, final IceInternal.EndpointI_connectors callback)
    {
        IceInternal.EndpointI_connectors cb = new IceInternal.EndpointI_connectors()
        {
            @Override
            public void connectors(java.util.List<IceInternal.Connector> connectors)
            {
                java.util.List<IceInternal.Connector> l = new java.util.ArrayList<IceInternal.Connector>();
                for(IceInternal.Connector c : connectors)
                {
                    l.add(new ConnectorI(_instance, c, _delegate.host(), _delegate.port(), _resource));
                }
                callback.connectors(l);
            }

            @Override
            public void exception(Ice.LocalException ex)
            {
                callback.exception(ex);
            }
        };
        _delegate.connectors_async(selType, cb);
    }

    @Override
    public IceInternal.Acceptor acceptor(IceInternal.EndpointIHolder endpoint, String adapterName)
    {
        IceInternal.EndpointIHolder delEndp = new IceInternal.EndpointIHolder();
        IceInternal.Acceptor delAcc = _delegate.acceptor(delEndp, adapterName);
        if(delEndp.value != null)
        {
            endpoint.value = new EndpointI(_instance, delEndp.value, _resource);
        }
        return new AcceptorI(_instance, delAcc);
    }

    @Override
    public java.util.List<IceInternal.EndpointI> expand()
    {
        java.util.List<IceInternal.EndpointI> endps = _delegate.expand();
        java.util.List<IceInternal.EndpointI> l = new java.util.ArrayList<IceInternal.EndpointI>();
        for(IceInternal.EndpointI e : endps)
        {
            l.add(e == _delegate ? this : new EndpointI(_instance, e, _resource));
        }
        return l;
    }

    @Override
    public boolean equivalent(IceInternal.EndpointI endpoint)
    {
        if(!(endpoint instanceof EndpointI))
        {
            return false;
        }
        EndpointI wsEndpointI = (EndpointI)endpoint;
        return _delegate.equivalent(wsEndpointI._delegate);
    }

    @Override
    synchronized public int hashCode()
    {
        int h = _delegate.hashCode();
        h = IceInternal.HashUtil.hashAdd(h, _resource);
        return h;
    }

    @Override
    public String options()
    {
        //
        // WARNING: Certain features, such as proxy validation in Glacier2,
        // depend on the format of proxy strings. Changes to toString() and
        // methods called to generate parts of the reference string could break
        // these features. Please review for all features that depend on the
        // format of proxyToString() before changing this and related code.
        //
        String s = _delegate.options();

        if(_resource != null && _resource.length() > 0)
        {
            s += " -r ";
            boolean addQuote = _resource.indexOf(':') != -1;
            if(addQuote)
            {
                s += "\"";
            }
            s += _resource;
            if(addQuote)
            {
                s += "\"";
            }
        }

        return s;
    }

    @Override
    public int compareTo(IceInternal.EndpointI obj) // From java.lang.Comparable
    {
        if(!(obj instanceof EndpointI))
        {
            return type() < obj.type() ? -1 : 1;
        }

        EndpointI p = (EndpointI)obj;
        if(this == p)
        {
            return 0;
        }

        int v = _resource.compareTo(p._resource);
        if(v != 0)
        {
            return v;
        }

        return _delegate.compareTo(p._delegate);
    }

    @Override
    protected boolean checkOption(String option, String argument, String endpoint)
    {
        switch(option.charAt(1))
        {
        case 'r':
        {
            if(argument == null)
            {
                throw new Ice.EndpointParseException("no argument provided for -r option in endpoint " + endpoint +
                    _delegate.options());
            }
            _resource = argument;
            return true;
        }

        default:
        {
            return false;
        }
        }
    }

    private Instance _instance;
    private IceInternal.IPEndpointI _delegate;
    private String _resource;
}
